src/mpi/comm/commutil.c

```diff
@@ -1110,6 +1112,28 @@ int MPIR_Comm_delete_internal(MPIR_Comm * comm_ptr)
      * communicator must not be freed.  That is the reason for the
      * test on mpi_errno here. */
     if (mpi_errno == MPI_SUCCESS) {
+        /* receive any unmatched messages to clear the queue and avoid context_id
+         * reuse issues. unmatched messages could be the result of user error,
+         * or send operations that were unable to be canceled by the device layer */
+        MPIR_Object_add_ref(comm_ptr);
+        int flag;
+        MPI_Status status;
+        do {
+            mpi_errno = MPID_Iprobe(MPI_ANY_SOURCE, MPI_ANY_TAG, comm_ptr, 0, &flag, &status);
+            MPIR_ERR_CHECK(mpi_errno);
+            if (flag) {
+                /* receive the message, ignore truncation errors */
+                MPIR_Request *request;
+                MPID_Recv(NULL, 0, MPI_DATATYPE_NULL, status.MPI_SOURCE, status.MPI_TAG, comm_ptr,
+                          0, MPI_STATUS_IGNORE, &request);
+                if (request != NULL) {
+                    MPIR_Wait(&request->handle, MPI_STATUS_IGNORE);
+                }
+                unmatched_messages++;
+            }
+        } while (flag);
+        MPIR_Object_release_ref(comm_ptr, &in_use);
+
```

Comment:
I think we should give the user an option to switch this off. Maybe only do this when error checking is enabled?

Reply:
Sure, will fix.

Reply:
Guarded by `#ifdef HAVE_ERROR_CHECKING` now.

Reply:
Do we need adjust the test to work for both cases?

Reply:
The testsuite configure script will disable the `test/mpi/errors` directory if it finds `--enable-error-checking=no` in the configuration options, so I don't think any modification is needed.